csharp/src/Ice/VectoredBuffer.cs

```diff
@@ -0,0 +1,549 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Ice
+{
+    /// <summary>Represents a position if a VectoredBuffer, the position is compose of
+    /// the segment index and the relative offset in the segment.</summary>
+    public struct BufferPosition
+    {
+        /// <summary>Creates a new position from the segment and offset values.</summary>
+        /// <param name="segment">The zero based index of the segment.</param>
+        /// <param name="offset">The offset into the segment.</param>
+        public BufferPosition(int segment, int offset)
+        {
+            Segment = segment;
+            Offset = offset;
+        }
+
+        /// <summary>The zero based index of the segment.</summary>
+        public int Segment;
+        /// <summary>The offset into the segment.</summary>
+        public int Offset;
+    }
+
+    /// <summary>A byte buffer class that holds data in a list of byte arrays,
+    /// the buffer expands its capacity by adding new arrays to the list,
+    /// this allow to grow without need to reallocate data.</summary>
+    public class VectoredBuffer
+    {
+        /// <summary>Number of bytes that the buffer can hold.</summary>
+        public int Capacity { get; private set; }
+
+        public BufferPosition Pos
+        {
+            get => _position;
+            set
+            {
+                int index = value.Segment;
+                _currentSegmentOffset = 0;
+                for (int i = 0; i < index; i++)
+                {
+                    _currentSegmentOffset += Segments[i].Count;
+                }
+                _currentSegment = Segments[index];
+
+                if (value.Offset > _currentSegment.Count)
+                {
+                    throw new ArgumentException("The position offset is invalid for the current buffer", nameof(value));
+                }
+                _position = value;
+            }
+        }
+
+        /// <summary>Number of bytes from current position to the end.</summary>
+        public int Remaining => Size - (_position.Offset + _currentSegmentOffset);
+
+        /// <summary>Determines the current size of the stream, this correspond
+        /// to the number of bytes already writen to the stream.</summary>
+        /// <value>The current size.</value>
+        public int Size { get; private set; }
+
+        /// <summary>List of array segments own by the buffer, the last segment size reflects the size
+        /// of the underlying array until Prepare is call after that point it reflects the portion
+        /// of the last segment that is in use.</summary>
+        public IList<ArraySegment<byte>> Segments { get; }
+        // The segment curretnly used by write operations
+        private ArraySegment<byte> _currentSegment;
+        // Cached offset of the current segment
```

Comment:
This `_currentSegmentOffset` apparently corresponds to the combined size of all segments prior to (but not including) the current segment. I find the name misleading - it suggests the offset in the current segment.
What about `_sizeOfPriorSegments`?